### PR TITLE
[InstCombine] Drop `samesign` in InstCombine

### LINKED ISSUE #112476: [InstCombine] `samesign` flag should be dropped in `InstCombinerImpl::foldICmpAndConstConst`
Reproducer:
```
define i1 @src(i32 %x, i32 %y) {
  %shf = lshr i32 %x, %y
  %or = or i32 %shf, %x
  %and = and i32 %or, 1
  %ret = icmp samesign ne i32 %and, 0
  ret i1 %ret
}

define i1 @tgt(i32 %x, i32 %y) {
  %shf1 = shl nuw i32 1, %y
  %or2 = or i32 %shf1, 1
  %and3 = and i32 %x, %or2
  %ret = icmp samesign ne i32 %and3, 0
  ret i1 %ret
}
```
```
Transformation doesn't verify!

ERROR: Target is more poisonous than source

Example:
i32 %x = #x80000000 (2147483648, -2147483648)
i32 %y = #x0000001f (31)

Source:
i32 %shf = #x00000001 (1)
i32 %or = #x80000001 (2147483649, -2147483647)
i32 %and = #x00000001 (1)
i1 %ret = #x1 (1)

Target:
i32 %shf1 = #x80000000 (2147483648, -2147483648)
i32 %or2 = #x80000001 (2147483649, -2147483647)
i32 %and3 = #x80000000 (2147483648, -2147483648)
i1 %ret = poison
Source value: #x1 (1)
Target value: poison
```

### PR DESCRIPTION
**Pull request by dtcxzyw**
Closes https://github.com/llvm/llvm-project/issues/112476.

### CODE REVIEW

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp [hunk 1]
```diff
@@ -1738,6 +1738,7 @@ Instruction *InstCombinerImpl::foldICmpAndShift(ICmpInst &Cmp,
 
     // Compute X & (C2 << Y).
     Value *NewAnd = Builder.CreateAnd(Shift->getOperand(0), NewShift);
+    Cmp.setSameSign(false);
     return replaceOperand(Cmp, 0, NewAnd);
```
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:1742]: Can we create a new icmp here instead? I don't see a good reason to do in-place modification here, and it causes this kind of issue...

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp [hunk 2]
```diff
@@ -1738,7 +1738,8 @@ Instruction *InstCombinerImpl::foldICmpAndShift(ICmpInst &Cmp,
 
     // Compute X & (C2 << Y).
     Value *NewAnd = Builder.CreateAnd(Shift->getOperand(0), NewShift);
-    return replaceOperand(Cmp, 0, NewAnd);
+    return ICmpInst::Create(Instruction::ICmp, Cmp.getPredicate(), NewAnd,
+                            Cmp.getOperand(1));
```
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:1742]:
```suggestion
    return new ICmpInst(Cmp.getPredicate(), NewAnd, Cmp.getOperand(1));
```

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp [hunk 3]
```diff
@@ -1844,7 +1845,8 @@ Instruction *InstCombinerImpl::foldICmpAndConstConst(ICmpInst &Cmp,
                                                /*HasNUW=*/true),
                              One, Or->getName());
         Value *NewAnd = Builder.CreateAnd(A, NewOr, And->getName());
-        return replaceOperand(Cmp, 0, NewAnd);
+        return ICmpInst::Create(Instruction::ICmp, Cmp.getPredicate(), NewAnd,
+                                Cmp.getOperand(1));
```
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:1849]:
```suggestion
        return new ICmpInst(Cmp.getPredicate(), NewAnd, Cmp.getOperand(1));
```

**Review by nikic — APPROVED** [review]
LGTM

**Review by antoniofrighetto — APPROVED** [review]

**Pull request merged**